feat: ignore mount dir check in csi node stage/publish

andyzhangx opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR ignores dir check in csi node stage/publish, as a proceeding fix as PR(https://github.com/kubernetes/kubernetes/pull/88569), csi driver should check target dir in node stage/publish func other than csi attacher/mounter.

There are many reasons why we should skip the dir check in csi attacher/mounter:
 - target mount dir could be in broken
 - on windows, we should not create dir in csi attacher/mounter since "bind mount" does not apply on windows
 - every csi driver should have its own logic to check whether target mount dir is correct or not

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #86784

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
ACTION REQUIRED: For CSI drivers, kubelet no longer creates the target_path for NodePublishVolume in accordance with the CSI spec. Kubelet also no longer checks if staging and target paths are mounts or corrupted. CSI drivers need to be idempotent and do any necessary mount verification.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

/assign @jsafrane @msau42 @davidz627  @gnufied
cc @seh @kubernetes/sig-storage-pr-reviews
/priority important-soon
/sig storage

andyzhangx committed:
fix: ignore dir check in csi node stage/publish

andyzhangx commented:
/test pull-kubernetes-e2e-kind

andyzhangx commented:
/test pull-kubernetes-e2e-kind-ipv6
/test pull-kubernetes-e2e-gce

seh commented on pkg/volume/csi/csi_mounter.go:
```diff
@@ -217,12 +201,6 @@ func (c *csiMountMgr) SetUpAt(dir string, mounterArgs volume.MounterArgs) error
 		return fmt.Errorf("volume source not found in volume.Spec")
 	}
 
-	// create target_dir before call to NodePublish
```
Thank you for the patch, @andyzhangx. I noticed that we used to call `os.MkdirAll` here, but now we won't. Can you clarify whether the CSI plugin should still find that this target directory already exists, due to it being created earlier, or are we transferring new responsibility to drivers to create this target directory?

andyzhangx replied:
I tend to create this target directory by driver itself, e.g. on windows, target directory is actually a symbolic link, if we create target directory here, then on Windows CSI driver, it needs to remove this target directory and create a symbolic link.
cc @kkmsft @ddebroy

seh replied:
I see. In my driver, and those that I've studied, we create the target path in the `NodePublishVolume` handler if the directory doesn't already exist, so I wouldn't mind this change in behavior.

jsafrane replied:
CSI says:

>   // The CO SHALL ensure that the parent directory of this path exists
  // and that the process serving the request has `read` and `write`
  // permissions to that parent directory.
  ...
  // Creation of target_path is the responsibility of the SP.

Should kubelet `MkdirAll()` the parent then?

andyzhangx replied:
fixed

jsafrane replied:
I meant, creation of `deviceMountPath` is responsiblity of a CSI driver, while kubelet should create just the parent, i.e. `MkdirAll(Dir(deviceMountPath))`.

Note that the change could break some CSI drivers that expected kubelet to create whole path. I'm not sure anyone tested with something else than Kubernetes... @msau42, what do you think?

andyzhangx replied:
@jsafrane o, that's what I did in the beginning, only create parent dir, and csi driver create `deviceMountPath`, and also I am not sure it would break sth. 
I reverted back the first commit and now this PR only checks the error when `MkdirAll(dir)`, it should be safer.

My initial opinion about this:
```
I tend to create this target directory by driver itself, e.g. on windows, target directory is actually a symbolic link, if we create target directory here, then on Windows CSI driver, it needs to remove this target directory and create a symbolic link.
```

msau42 replied:
We have an issue for this already and yes, it's a problem for windows: https://github.com/kubernetes/kubernetes/issues/75535

I am ok with removing creation of mount path, but let's call it out in the release notes so that drivers that were uncompliant have a chance to fix it

kkmsft replied:
I had similar thought while getting the azuredisk csi driver to work in windows. But there are other drivers (eg. https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/master/pkg/secrets-store) which assumes the directory existance. Hence choose to work around this issue in the windows azure disk csi. My recommendation is that we hold off until 1.18 release rolls out. Make the change early in 1.19 and inform folks via sig storage channels - slack, mailing list , meetings about this giving opportunity to work the changes out by the time 1.19 comes out. Thoughts ?

msau42 replied:
Sounds good. How about we at least make a deprecation notice this release?

msau42 replied:
Are we mainly concerned with removal of mount path creation?

Do we want to proceed with removal of mount point check?

kkmsft replied:
> Sounds good. How about we at least make a deprecation notice this release?

Yes, SGTM 👍

kkmsft replied:
I think we should put in all three notes in deprecation notice 

1. Create if not present of the mount directory, 
2. Check whether its mounted 
3. Corruption check 

all getting deprecated in the release note and proceed with the removal at the beginning of the 1.19 cycle.  Although I don’t know of which drivers would break in the (2) and (3), it looks very easy possibility to break things late the cycle with this.

feiskyer commented:
/milestone v1.18

andyzhangx commented:
/test pull-kubernetes-e2e-gce-csi-serial

andyzhangx commented:
/test pull-kubernetes-e2e-kind-ipv6
/test pull-kubernetes-e2e-gce
/test pull-kubernetes-e2e-kind

jsafrane commented:
/test pull-kubernetes-e2e-gce

jsafrane commented on pkg/volume/csi/csi_attacher.go:
```diff
@@ -293,11 +276,11 @@ func (c *csiAttacher) MountDevice(spec *volume.Spec, devicePath string, deviceMo
-	}
-	klog.V(4).Info(log("created target path successfully [%s]", deviceMountPath))
 	dataDir := filepath.Dir(deviceMountPath)
+	if err = os.MkdirAll(dataDir, 0750); err != nil {
```
`deviceMountPath` is used as `NodeStageVolumeRequest.staging_target_path`. CSI says:

>  // The CO SHALL ensure
  // that the path is directory and that the process serving the
  // request has `read` and `write` permission to that directory. The
  // CO SHALL be responsible for creating the directory if it does not
  // exist.
  string staging_target_path = 3;

So, Kubernetes should `MkDir` whole `deviceMountPath`. Or CSI needs to be fixed. Or worked around, as we do with block volumes - CSI drivers stage block devices as *files* inside `staging_target_path` *directory*).

gnufied replied:
are these addressed yet?

jsafrane commented on pkg/volume/csi/csi_attacher.go:
```diff
@@ -315,7 +298,7 @@ func (c *csiAttacher) MountDevice(spec *volume.Spec, devicePath string, deviceMo
 			// clean up metadata
 			klog.Errorf(log("attacher.MountDevice failed: %v", err))
-			if err := removeMountDir(c.plugin, deviceMountPath); err != nil {
+			if err := removeMountDir(c.plugin, dataDir); err != nil {
```
Again, whole deviceMountPath should be removed.

gnufied replied:
this needs to be addressed still..

jsafrane replied:
This was fixed.

jsafrane reviewed (commented):
I like removing the mount check, however, I am not so sure about removing `Mkdir`s - it needs to be done very carefully.

cc @msau42 @davidz627 @gnufied

andyzhangx committed:
fix comments

andyzhangx reviewed (commented):
@jsafrane addressed your comments, PTAL, thanks.

andyzhangx commented:
/test pull-kubernetes-e2e-gce-storage-snapshot 
/test pull-kubernetes-e2e-kind-ipv6

andyzhangx commented on pkg/volume/csi/csi_attacher.go:
```diff
@@ -228,23 +228,6 @@ func (c *csiAttacher) MountDevice(spec *volume.Spec, devicePath string, deviceMo
-		}
-	}
-
-	if mounted && !corruptedDir {
```
I realized this could be a breaking change for lots of CSI drivers, hold for discussion
/hold

kkmsft replied:
If the directory is already mounted and not corrupt  shouldn't we return successfully ?

kkmsft replied:
Okay, got it - we offload all of the path related checks (of the mount directories only) to the CSI drivers ?

smourapina commented:
Hello @andyzhangx, @jsafrane!
Bug Triage team here. We are now in code freeze for release 1.18, so I will go ahead reset the milestone to 1.19. Please let me know in case you have questions!
/milestone v1.19

msau42 commented:
/hold cancel
let's try to get this early in 1.19 cycle to get soak time

msau42 commented on pkg/volume/csi/csi_mounter.go:
```diff
@@ -218,8 +202,13 @@ func (c *csiMountMgr) SetUpAt(dir string, mounterArgs volume.MounterArgs) error
 	// create target_dir before call to NodePublish
-	if err := os.MkdirAll(dir, 0750); err != nil && !corruptedDir {
-		return errors.New(log("mounter.SetUpAt failed to create dir %#v:  %v", dir, err))
+	if err := os.MkdirAll(dir, 0750); err != nil {
```
We should Mkdir the parent(dir) to align with the csi spec wording. I believe then we don't need the corruptedDir check because the parent dir is out of the driver's control

msau42 commented:
Can you update the release note to say something like:
```
ACTION REQUIRED: For CSI drivers, kubelet no longer creates the target_path for NodePublishVolume in accordance with the CSI spec. Kubelet also no longer checks if staging and target paths are mounts or corrupted. CSI drivers need to be idempotent and do any necessary mount verification.
```

andyzhangx committed:
fix: comments(only create parent dir)

andyzhangx commented:
pls take a look at this PR again, now only create parent in node stage/publish

andyzhangx commented:
/test pull-kubernetes-e2e-kind
/test pull-kubernetes-e2e-kind-ipv6
/test pull-kubernetes-e2e-gce

andyzhangx commented:
/test pull-kubernetes-e2e-kind
/test pull-kubernetes-e2e-kind-ipv6
/test pull-kubernetes-e2e-gce
/test pull-kubernetes-e2e-gce-storage-slow

kbruner commented:
Hello from the bug triage team! This issue has not been updated for a long time, so I'd like to check what's the status. The code freeze is starting June 25th (about 4 from now) and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.

As the PR is tagged for 1.19, is it still planned for this release?

msau42 commented:
The release team has asked all sigs to minimize the amount of potentially disruptive changes in the 1.19 cycle. Given that this change has the potential to break csi drivers that were not idempotent or relied on K8s behavior to create the mount directory, I think it's safer to defer this to 1.20. And in the 1.19 cycle, we continue to publicize to CSI vendors that this change is coming.

kbruner commented:
@msau42 I will go ahead and move this to milestone 1.20. Thanks!

kbruner commented:
/milestone 1.20

kbruner commented:
/milestone v1.20

derp

msau42 commented:
/assign @jsafrane

jsafrane commented:
Finally got some time to test this patch with a real driver (AWS EBS). `NodePublish`/`Unpublish` changes look OK.

`NodeStage` behaves differently. When two pods that use the same volume run on a single node:
* Without this PR, I can see one `NodeStage` and two `NodePublishes`, as expected.
* With this PR, I can see `NodeStage` + `NodePublish` for the first pod and separate `NodeStage` + `NodePublish` for the second pod.

In both cases, `NodeUnstage` is called only once.

I can get a single `NodeStage` with a small change here: https://github.com/kubernetes/kubernetes/blob/a6d8e6c5c2e5e25a0dd6a003497519a226128161/pkg/volume/util/operationexecutor/operation_generator.go#L575

-> 
```
if volumeDeviceMounter != nil && actualStateOfWorld.GetDeviceMountState(volumeToMount.VolumeName) != DeviceGloballyMounted {
```
I.e. call `MountDevice` only if the device is not mounted yet (it's the first call or previous `MountDevice` did not succeed).

This changes `nodeExpandVolume` a bit - if it fails after `MountDevice` and before `SetUp`, the device is already marked as DeviceMounted. On the next retry, Kubelet will instead call `SetUp` / `NodePublish` and try `nodeExpandVolume` after that. cc @gnufied to check if it matters.

gnufied commented on pkg/volume/csi/csi_attacher.go:
```diff
@@ -293,11 +276,12 @@ func (c *csiAttacher) MountDevice(spec *volume.Spec, devicePath string, deviceMo
-	klog.V(4).Info(log("created target path successfully [%s]", deviceMountPath))
-	dataDir := filepath.Dir(deviceMountPath)
+	klog.V(4).Info(log("created target path successfully [%s]", parentDir))
+	dataDir := parentDir
```
@andyzhangx @msau42 I think we have to create staging path at least, to be spec complaint.

gnufied commented:
@jsafrane yeah volume expansion is bit of a concern here. The change you are proposing will cause NodeExpandVolume to be not retried. Officially - calling of `NodeExpandVolume` between stage and publish is being deprecated and will be removed in 1.21 I think but until that happens, I would prefer we did not change the behaviour.

jsafrane commented:
> calling of NodeExpandVolume between stage and publish is being deprecated and will be removed in 1.21 I think but until that happens, I would prefer we did not change the behaviour.

Ok, so what can we do is to mark the volume as "uncertain" after `nodeExpandVolume` error here: https://github.com/kubernetes/kubernetes/blob/7594702b22e57b7ecd029e19e7453480cee60064/pkg/volume/util/operationexecutor/operation_generator.go#L614-L619
This way, each retry will call `MountDevice`( = `NodeStageVolume`) and `nodeExpandVolume` after that, abusing `MountDevice` a bit. But it should be idempotent, right? :-)

@andyzhangx, can you please fix the PR with this + https://github.com/kubernetes/kubernetes/pull/88759#issuecomment-704204620?

andyzhangx commented:
> > calling of NodeExpandVolume between stage and publish is being deprecated and will be removed in 1.21 I think but until that happens, I would prefer we did not change the behaviour.
> 
> Ok, so what can we do is to mark the volume as "uncertain" after `nodeExpandVolume` error here:
> 
> https://github.com/kubernetes/kubernetes/blob/7594702b22e57b7ecd029e19e7453480cee60064/pkg/volume/util/operationexecutor/operation_generator.go#L614-L619
> 
> 
> This way, each retry will call `MountDevice`( = `NodeStageVolume`) and `nodeExpandVolume` after that, abusing `MountDevice` a bit. But it should be idempotent, right? :-)
> @andyzhangx, can you please fix the PR with this + [#88759 (comment)](https://github.com/kubernetes/kubernetes/pull/88759#issuecomment-704204620)?

sure, I am on vacation this week, will take a look next week.

bai commented:
Hey 👋 Bug Triage here. Just wanted to post a friendly reminder that the code freeze is starting 12.11.2020 (about 3-4 weeks from now) and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.

As the PR is tagged for 1.20, is it still planned for this release?

andyzhangx commented:
@jsafrane I still don't have time to continue on this PR this week, @jsafrane are you willing to work on this? I have added you as collaborator  to my k8s repo. thanks.

jsafrane committed:
Restore staging path creation

CSI says about NodeStage (=MountDevice):
// The CO SHALL ensure [...]
// that the path is directory and that the process serving the
// request has `read` and `write` permission to that directory. The
// CO SHALL be responsible for creating the directory if it does not
// exist.

jsafrane committed:
Mark MountDevice as uncertain after failed resize

When NodeExpand fails between MountDevice and SetUp, make sure the next
NodeExpand attempt is called before SetUp.

jsafrane committed:
Call MountDevice only once

Cann MountDevice only when the volume was not device-mounted before.

jsafrane commented:
Added three commits fixing remaining issues with this PR:

* Kubelet now creates directory for NodeStage, as required by CSI spec.
* When NodeExpand fails, corresponding MountDevice is marked as uncertain to retry NodeExpand again (and call unnecessary MountDevice again too, but we don't track "undertain-only-resize-pendig" state).
* Call MountDevice only when the device is not device-mounted yet, so a volume used in two pods on a single node gets only once MountDevice / NodeStage + 2x SetUp / NodePublish.

jsafrane commented:
And tested a bit on AWS.

msau42 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -568,7 +568,7 @@ func (og *operationGenerator) GenerateMountVolumeFunc(
 		}
 
-		if volumeDeviceMounter != nil {
+		if volumeDeviceMounter != nil && actualStateOfWorld.GetDeviceMountState(volumeToMount.VolumeName) != DeviceGloballyMounted {
```
Is it possible to add unit tests for these changes?

jsafrane replied:
operation_generator.go does not have a big unit test coverage, I'll try to test it manually with a mock driver + some error injection.

andyzhangx replied:
@jsafrane we have azure file csi driver that has implemented `NodeStageVolume`, and for this code change, it only call one `NodeStageVolume`, it would not work if that volume is going to mount on multiple nodes, shall we revert this code change? find details here: https://github.com/kubernetes-sigs/azurefile-csi-driver/issues/620

jsafrane replied:
I am not sure I understand. The code above makes sure that `NodeStageVolume` should be called only once on each node where a shared volume is used. The volume can still be used in multiple nodes.

andyzhangx replied:
from what I observe, there is a little possibility that `NodeStageVolume` is not invoked, only invoke `NodePublishVolume`, and it only happens from k8s v1.20.x, do you know what could be the root cause? thanks.
```
I0502 14:51:38.501550       1 utils.go:118] GRPC call: /csi.v1.Node/NodePublishVolume
I0502 14:51:38.501574       1 utils.go:119] GRPC request: {"staging_target_path":"/var/lib/kubelet/plugins/kubernetes.io/csi/pv/azurefile-4885-file.csi.azure.com-preprovsioned-pv-ql6t7/globalmount","target_path":"/var/lib/kubelet/pods/ed800670-69de-4a1e-902c-e0ceeca4ec68/volumes/kubernetes.io~csi/azurefile-4885-file.csi.azure.com-preprovsioned-pv-ql6t7/mount","volume_capability":{"AccessType":{"Mount":{}},"access_mode":{"mode":1}},"volume_context":{"csi.storage.k8s.io/ephemeral":"false","csi.storage.k8s.io/pod.name":"azurefile-volume-tester-2tpth","csi.storage.k8s.io/pod.namespace":"azurefile-4885","csi.storage.k8s.io/pod.uid":"ed800670-69de-4a1e-902c-e0ceeca4ec68","csi.storage.k8s.io/serviceAccount.name":"default"},"volume_id":"kubetest-u33fnt21#f7bf7d45cebcf4286987d85#pre-provisioned-multiple-pods#"}
I0502 14:51:38.501732       1 nodeserver.go:84] NodePublishVolume: mounting /var/lib/kubelet/plugins/kubernetes.io/csi/pv/azurefile-4885-file.csi.azure.com-preprovsioned-pv-ql6t7/globalmount at /var/lib/kubelet/pods/ed800670-69de-4a1e-902c-e0ceeca4ec68/volumes/kubernetes.io~csi/azurefile-4885-file.csi.azure.com-preprovsioned-pv-ql6t7/mount with mountOptions: [bind]
I0502 14:51:38.501750       1 mount_linux.go:175] Mounting cmd (mount) with arguments ( -o bind /var/lib/kubelet/plugins/kubernetes.io/csi/pv/azurefile-4885-file.csi.azure.com-preprovsioned-pv-ql6t7/globalmount /var/lib/kubelet/pods/ed800670-69de-4a1e-902c-e0ceeca4ec68/volumes/kubernetes.io~csi/azurefile-4885-file.csi.azure.com-preprovsioned-pv-ql6t7/mount)
E0502 14:51:38.503747       1 mount_linux.go:179] Mount failed: exit status 32
Mounting command: mount
Mounting arguments:  -o bind /var/lib/kubelet/plugins/kubernetes.io/csi/pv/azurefile-4885-file.csi.azure.com-preprovsioned-pv-ql6t7/globalmount /var/lib/kubelet/pods/ed800670-69de-4a1e-902c-e0ceeca4ec68/volumes/kubernetes.io~csi/azurefile-4885-file.csi.azure.com-preprovsioned-pv-ql6t7/mount
Output: mount: /var/lib/kubelet/pods/ed800670-69de-4a1e-902c-e0ceeca4ec68/volumes/kubernetes.io~csi/azurefile-4885-file.csi.azure.com-preprovsioned-pv-ql6t7/mount: special device /var/lib/kubelet/plugins/kubernetes.io/csi/pv/azurefile-4885-file.csi.azure.com-preprovsioned-pv-ql6t7/globalmount does not exist.

E0502 14:51:38.503833       1 utils.go:123] GRPC error: rpc error: code = Internal desc = Could not mount "/var/lib/kubelet/plugins/kubernetes.io/csi/pv/azurefile-4885-file.csi.azure.com-preprovsioned-pv-ql6t7/globalmount" at "/var/lib/kubelet/pods/ed800670-69de-4a1e-902c-e0ceeca4ec68/volumes/kubernetes.io~csi/azurefile-4885-file.csi.azure.com-preprovsioned-pv-ql6t7/mount": mount failed: exit status 32
Mounting command: mount
Mounting arguments:  -o bind /var/lib/kubelet/plugins/kubernetes.io/csi/pv/azurefile-4885-file.csi.azure.com-preprovsioned-pv-ql6t7/globalmount /var/lib/kubelet/pods/ed800670-69de-4a1e-902c-e0ceeca4ec68/volumes/kubernetes.io~csi/azurefile-4885-file.csi.azure.com-preprovsioned-pv-ql6t7/mount
Output: mount: /var/lib/kubelet/pods/ed800670-69de-4a1e-902c-e0ceeca4ec68/volumes/kubernetes.io~csi/azurefile-4885-file.csi.azure.com-preprovsioned-pv-ql6t7/mount: special device /var/lib/kubelet/plugins/kubernetes.io/csi/pv/azurefile-4885-file.csi.azure.com-preprovsioned-pv-ql6t7/globalmount does not exist.
```

andyzhangx replied:
I think I found the issue, there are multiple pods referencing same volume in the test case, those pods would be created, deleted in serial, so usually it won't have any issue, every pod creation and termination would have corresponding `NodeStageVolume` and then `NodePublishVolume` one by one; the error case could be `NodeUnstageVolume` takes some time(delay), and for next pod, it could be still in `DeviceGloballyMounted`, and then skip `NodeStageVolume` and after a while, the former `NodeUnstageVolume` completed, so for next pod, it only has `NodePublishVolume` func, is that possible @jsafrane

andyzhangx replied:
another common error case: if two csi driver PVs has the same `volumeHandle` value, `NodeStageVolume` will only happen once and the same PV mount won't have `NodeStageVolume`, it would fail.

jsafrane replied:
If they have the same `volumeHandle`, they're the same volumes and hence a single NodeStage is called. From CSI `volume_id` description:
```
  // This field MUST contain enough information to uniquely identify
  // this specific volume vs all other volumes supported by this plugin.
  // This field SHALL be used by the CO in subsequent calls to refer to
  // this volume.
  // The SP is NOT responsible for global uniqueness of volume_id across
  // multiple SPs.
  string volume_id = 2;
```

jsafrane replied:
> `NodeUnstageVolume` takes some time(delay), and for next pod, it could be still in `DeviceGloballyMounted`, and then skip `NodeStageVolume` and after a while, the former `NodeUnstageVolume` completed, so for next pod, it only has `NodePublishVolume` func, is that possible @jsafrane

It should not be possible. Kubelet should run only a single operator per volume (MountVolume or UnmountVolume). The aforementioned behavior would be a bug, if it really happens.

rohitChaware replied:
This affects the following scenario, can you please share your thoughts:

If there is a pod(let's say part of a StatefulSet) running on a node which uses a volume and the pod is deleted, the `NodeUnstageVolume` RPC gets called.

If completion of this `NodeUnstageVolume` RPC call takes time(for example: due to a wait on completion on `umount` Linux command), for the new replacement pod(scheduled on the same node) kubelet will find the volume as `DeviceGloballyMounted` and calls `NodePublishVolume` RPC without calling `NodeStageVolume` RPC. 

Later the `NodeUnstageVolume` RPC's unmount operation which was in progress would complete and unmount the volume. Thus, effectively for the new pod the staging doesn't happen.

andyzhangx replied:
@rohitChaware did you hit such race condition issue recently?

rohitChaware replied:
Yes @andyzhangx , reproduced the above scenario on K8s version 1.20.6. 

`NodeStageVolume` RPC isn't getting called in case when `NodeUnstageVolume` operation takes longer to complete(more than 2 minutes) and a new pod using the same volume is scheduled on the same node.

The flow of RPC calls go as expected for versions < 1.20.

msau42 replied:
https://github.com/kubernetes/kubernetes/pull/100183 may help

msau42 commented on pkg/volume/csi/csi_mounter.go:
```diff
@@ -218,10 +202,11 @@ func (c *csiMountMgr) SetUpAt(dir string, mounterArgs volume.MounterArgs) error
 	// create target_dir before call to NodePublish
-	if err := os.MkdirAll(dir, 0750); err != nil && !corruptedDir {
-		return errors.New(log("mounter.SetUpAt failed to create dir %#v:  %v", dir, err))
+	parentDir := filepath.Dir(dir)
```
Can we add a unit test to make sure we're creating the right directories?

jsafrane replied:
Added

jsafrane committed:
Add unit test for staging path creation

msau42 commented:
/lgtm
/approve

jingxu97 commented:
This PR breaks gce pd driver test for Windows due to a bug in the driver side.
Seems currently there is no good way to monitor changes in k/k which might affect csi driver functioning ...

msau42 commented:
We discussed this in the csi meeting. We already put out a deprecation notice with 1.18, and the sanity test in csi-test has not been creating the parent directory for almost a year now, so plugins should have had enough time to fix their drivers.

We'll send out an email reminder to storage/csi lists to warn driver authors that this change is coming, and they can use csi-test to validate their driver can handle this properly.

For Windows testing, I think we need to get the CI in a good enough state that we can run windows csi tests as a presubmit to catch problems like this earlier. And also look into how to support csi sanity on Windows.

jingxu97 commented:
I catch this issue through windows ci tests for pd driver. A problem is we cannot run csi driver presubmit test in k/k changes. CI tests can help to catch issues afterwards.

msau42 commented:
We have some csi drivers defined in k/k e2es: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/drivers/csi.go

And we have an optional presubmit job to run it whenever a volume file is changed: pull-kubernetes-e2e-gce-csi-serial 

I we can stabilize the windows presubmit, I think we can add a similar job.

jingxu97 commented:
got it, thanks! will check and add it for Windows.
